File: Sources/PolyReadCore/FrameMath.swift
import Foundation

/// §3 — "one frame is 600 samples at 24 kHz, so 40 frames/sec, 0.025 s/frame."
///
/// `duration` comes out of KokoroProsody unrounded. Round to at least 1 *before*
/// the gather and use the rounded values for timing, so the audio the acoustic
/// model emits and the timeline the scrubber reads agree exactly. Every rounding
/// in the codebase goes through `roundedFrames` so there is one rule, not five.
public enum FrameMath {
    public static let sampleRate: Double = 24_000
    public static let samplesPerFrame: Int = 600
    public static let framesPerSecond: Double = sampleRate / Double(samplesPerFrame)  // 40
    public static let secondsPerFrame: Double = Double(samplesPerFrame) / sampleRate  // 0.025

    /// The single rounding rule. `max(1, ...)` because a zero-frame token would
    /// collapse a word to zero width on the timeline and the highlight would skip it.
    public static func roundedFrames(_ raw: Float) -> Int {
        max(1, Int(raw.rounded()))
    }

    public static func roundedFrames(_ raw: [Float]) -> [Int] {
        raw.map(roundedFrames)
    }

    public static func seconds(frames: Int) -> TimeInterval {
        Double(frames) * secondsPerFrame
    }

    public static func frames(seconds: TimeInterval) -> Int {
        Int((seconds * framesPerSecond).rounded())
    }

    public static func samples(frames: Int) -> Int {
        frames * samplesPerFrame
    }

    public static func seconds(samples: Int) -> TimeInterval {
        Double(samples) / sampleRate
    }

    public static func samples(seconds: TimeInterval) -> Int {
        Int((seconds * sampleRate).rounded())
    }
}

/// §5 / §4.6 — inter-chunk silence. Kokoro ignores inline markup, so paragraph
/// pauses are real silence inserted between rendered chunks.
public enum Pause {
    public static let paragraph: TimeInterval = 0.400   // §5: 300–500 ms
    public static let heading: TimeInterval = 0.400     // §4.6: 400 ms before and after
    public static let withinBlock: TimeInterval = 0     // a split block is one utterance

    public static func frames(_ interval: TimeInterval) -> Int {
        FrameMath.frames(seconds: interval)
    }
}

File: web/src/core/frameMath.ts
/**
 * §3 — "one frame is 600 samples at 24 kHz, so 40 frames/sec, 0.025 s/frame."
 *
 * Kokoro's duration predictor emits frames, unrounded. Round to at least 1
 * *before* the gather and use the rounded values for timing, so the audio the
 * decoder emits and the timeline the scrubber reads agree exactly. Every
 * rounding in the codebase goes through `roundedFrames`, so there is one rule
 * rather than five.
 */
export const SAMPLE_RATE = 24_000;
export const SAMPLES_PER_FRAME = 600;
export const FRAMES_PER_SECOND = SAMPLE_RATE / SAMPLES_PER_FRAME; // 40
export const SECONDS_PER_FRAME = SAMPLES_PER_FRAME / SAMPLE_RATE; // 0.025

/**
 * The single rounding rule. `max(1, ...)` because a zero-frame token would
 * collapse a word to zero width on the timeline and the highlight would skip
 * straight past it.
 */
export function roundedFrames(raw: number): number {
  return Math.max(1, Math.round(raw));
}

export function roundedFramesAll(raw: ArrayLike<number>): number[] {
  const out = new Array<number>(raw.length);
  for (let i = 0; i < raw.length; i++) out[i] = roundedFrames(raw[i]);
  return out;
}

export const secondsFromFrames = (frames: number): number => frames * SECONDS_PER_FRAME;
export const framesFromSeconds = (seconds: number): number => Math.round(seconds * FRAMES_PER_SECOND);
export const samplesFromFrames = (frames: number): number => frames * SAMPLES_PER_FRAME;
export const secondsFromSamples = (samples: number): number => samples / SAMPLE_RATE;
export const samplesFromSeconds = (seconds: number): number => Math.round(seconds * SAMPLE_RATE);

/**
 * §5 / §4.6 — inter-chunk silence. Kokoro ignores inline markup, so paragraph
 * pauses are real silence written between rendered chunks.
 */
export const Pause = {
  /** §5: 300-500 ms */
  paragraph: 0.4,
  /** §4.6: 400 ms before and after a heading */
  heading: 0.4,
  /** a split block is one utterance */
  withinBlock: 0,
  frames: (interval: number): number => framesFromSeconds(interval),
} as const;
